cocos/audio/android/CCThreadPool.h

```diff
@@ -0,0 +1,239 @@
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ 
+ Inspired by https://github.com/vit-vit/CTPL
+ 
+ ****************************************************************************/
+#ifndef __CCTHREAD_POOL_H__
+#define __CCTHREAD_POOL_H__
+
+//#include "platform/CCPlatformMacros.h"
+
+#include <functional>
+#include <memory>
+#include <thread>
+#include <queue>
+#include <mutex>
+#include <condition_variable>
+#include <vector>
+#include <atomic>
+#include <mutex>
+
+namespace cocos2d { namespace experimental {
+
+/**
+ * @addtogroup base
+ * @{
+ */
+
+class ThreadPool
+{
+public:
+
+    enum TaskType
+    {
+        TASK_TYPE_DEFAULT = 0,
+        TASK_TYPE_NETWORK,
+        TASK_TYPE_IO,
+        TASK_TYPE_AUDIO,
+        TASK_TYPE_USER = 1000
+    };
+
+    /*
+     * Gets the default thread pool which is a cached thread pool with default parameters.
+     */
+    static ThreadPool *getDefaultThreadPool();
+
+    /*
+     * Destroys the default thread pool
+     */
+    static void destroyDefaultThreadPool();
+
+    /*
+     * Creates a cached thread pool
+     * @note The return value has to be delete while it doesn't needed
+     */
+    static ThreadPool *newCachedThreadPool(int minThreadNum, int maxThreadNum, int shrinkInterval,
+                                           int shrinkStep, int stretchStep);
+
+    /*
+     * Creates a thread pool with fixed thread count
+     * @note The return value has to be delete while it doesn't needed
+     */
+    static ThreadPool *newFixedThreadPool(int threadNum);
+
+    /*
+     * Creates a thread pool with only one thread in the pool, it could be used to exeute multipy tasks serially in just one thread.
+     * @note The return value has to be delete while it doesn't needed
+     */
+    static ThreadPool *newSingleThreadPool();
+
+    // the destructor waits for all the functions in the queue to be finished
+    ~ThreadPool();
+
+    /* Pushs a task to thread pool
+     *  @param runnable The callback of the task executed in sub thread
+     *  @param type The task type, it's TASK_TYPE_DEFAULT if this argument isn't assigned
+     *  @note This function has to be invoked in cocos thread
+     */
+    void pushTask(const std::function<void(int)>& runnable, int type = TASK_TYPE_DEFAULT);
+
+    // Stops all tasks, it will remove all tasks in queue
+    void stopAllTasks();
+
+    // Stops some tasks by type
+    void stopTasksByType(int type);
+
+    // Gets the minimum thread numbers
+    inline int getMinThreadNum()
+    { return _minThreadNum; };
+
+    // Gets the maximum thread numbers
+    inline int getMaxThreadNum()
+    { return _maxThreadNum; };
+
+    // Gets the number of idle threads
+    int getIdleThreadNum();
+
+    // Gets the number of initialized threads
+    inline int getInitedThreadNum()
+    { return _initedThreadNum; };
+
+    // Gets the task number
+    size_t getTaskNum();
```

Comment:
it is better to add `const` for get functions. And it is better to use `int` or `size_t` for all get functions to make consistent.